contracts/orderbook/OrderBook.sol

```diff
@@ -0,0 +1,325 @@
+pragma solidity ^0.8.4;
+import "./OrderBookEvents.sol";
+import "./bZxInterfaces/IPriceFeeds.sol";
+import "./bZxInterfaces/ILoanToken.sol";
+import "./bZxInterfaces/IBZX.sol";
+import "./OrderBookStorage.sol";
+import "./dexSwaps.sol";
+
+interface UniswapFactory{
+	function getPair(address tokenA, address tokenB) external view returns(address pair);
+}
+interface UniswapPair{
+	function token0() external view returns(address);
+	function token1() external view returns(address);
+	function getReserves() external view returns(uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast);
+}
+contract OrderBook is OrderBookEvents,OrderBookStorage{
+	modifier onlyOwner(){
+		require(msg.sender == owner);_;
+	}
+    function executeTradeOpen(address trader, uint orderID, address keeper,address usedToken) internal returns(uint success){
+		
+		_safeTransferFrom(usedToken,trader,address(this),IERC(usedToken).balanceOf(trader),"");
+		(bool result, bytes memory data) = HistoricalOrders[trader][orderID].iToken.call(abi.encodeWithSignature("marginTrade(bytes32,uint256,uint256,uint256,address,address,bytes)",HistoricalOrders[trader][orderID].loanID,HistoricalOrders[trader][orderID].leverage,HistoricalOrders[trader][orderID].loanTokenAmount,HistoricalOrders[trader][orderID].collateralTokenAmount,HistoricalOrders[trader][orderID].base,address(this),HistoricalOrders[trader][orderID].loanData));

```

Comment:
to have a better type safety instead of `abi.encodeWithSignature("marginTrade` you can use `iToken(addres).marginTrade.selector`

Reply:
if you read `HistoricalOrders[trader][orderID]` once into memory to the struct. then it should clear out stack to deep error

Reply:
Thanks. This has been implemented.